components/script/dom/blob.rs

```diff
@@ -121,23 +155,53 @@ impl Blob {
             Some(blobparts) => blob_parts_to_bytes(blobparts),
         };
 
-        let slice = DataSlice::new(Arc::new(bytes), None, None);
-        Ok(Blob::new(global, slice, &blobPropertyBag.get_typestring()))
+        let slice = DataSlice::from_bytes(bytes);
+        Ok(Blob::new(global, BlobImpl::new_from_slice(slice), &blobPropertyBag.get_typestring()))
     }
 
-    pub fn get_data(&self) -> &DataSlice {
-        &self.data
+    /// Get a slice to inner data, this might incur synchronous read and caching
+    pub fn get_slice(&self) -> DataSlice {
+        match self.blob_impl {
+            BlobImpl::File(ref id, ref slice) => {
+                match *slice.borrow() {
+                    Some(ref s) => s.clone(),
+                    None => {
+                        let global = self.global();
+                        let s = read_file(global.r(), id.clone());
+                        *slice.borrow_mut() = Some(s.clone()); // Cached
+                        s
+                    }
+                }
+            }
+            BlobImpl::Memory(ref s) => s.clone()
+        }
     }
 }
 
+fn read_file(global: GlobalRef, id: Uuid) -> DataSlice {
+    // FIXME: After resource thread #11189 PR is landed
+    // let global = global.filemanager_thread();
+
+    let file_manager: IpcSender<FileManagerThreadMsg> = unimplemented!();
+    let (chan, recv) = ipc::channel().unwrap();
+    let _ = file_manager.send(FileManagerThreadMsg::ReadFile(chan, id));
+    let result = recv.recv().unwrap();
```

Comment:
I think `read_file` can be made `Option<DataSlice>` and we can throw a script-level exception or something at the dom API where it is used (like `Slice`)